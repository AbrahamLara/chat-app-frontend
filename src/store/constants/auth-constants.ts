export const LOGIN_USER = 'loginUser';
export const REGISTER_USER = 'registerUser';
export const SET_IS_AUTHENTICATED = 'setIsAuthenticated';
export const GET_USER_PROFILE = 'getUserProfile';
export const SET_USER_PROFILE = 'setUserProfile';
