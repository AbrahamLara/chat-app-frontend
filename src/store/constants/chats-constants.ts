export const SET_CHATS_LIST = 'setChatsList';
export const FETCH_CHATS = 'fetchChats';
export const CREATE_CHAT = 'createChat';
export const ADD_CHAT = 'addChat';
