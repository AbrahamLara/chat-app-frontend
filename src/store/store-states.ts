import { THEME } from '@/utils/theme-utils';
import { FormAlertMessage, AlertMessage } from '@/utils/alerts-utils';
import { UserChatItem } from '@/utils/chat-utils';

/**
 * Describes the state of application.
 */
interface AppState {
  /**
   * The current theme to render the app in. Either dark or light.
   */
  theme: THEME;

  /**
   * The list of chat groups the user is in.
   */
  chats?: ChatsState;

  /**
   * The user's authenticated state.
   */
  auth?: UserState;
}

interface ChatsState extends AlertsState {
  /**
   * Describes the state of the user's chat list.
   */
  chats: UserChatItem[];
}

/**
 * Describes the user's authenticated state.
 */
interface UserState extends AlertsState {
  isAuthenticated: boolean;
}

/**
 * Describes the state of alert types on the page.
 */
interface AlertsState {
  alerts: {
    /**
     * Login form error messages.
     */
    errors: (AlertMessage | FormAlertMessage)[];

    /**
     * Register form error messages.
     */
    successes: AlertMessage[];
  };
}

export { AppState, ChatsState, AlertsState, UserState };
