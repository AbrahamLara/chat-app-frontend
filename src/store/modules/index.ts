import { ModuleTree } from 'vuex';
import { RootState } from '@/store/store-states';
import alertsModule from '@/store/modules/alerts-module';

export const ALERTS_NAMESPACE = 'alerts';

const appModules: ModuleTree<RootState> = {
  [ALERTS_NAMESPACE]: alertsModule,
};

/**
 * This function namespaces the provided mutation name and returns it.
 */
export function namespaceAlerts(mutation: string) {
  return `${ALERTS_NAMESPACE}/${mutation}`;
}

export default appModules;
