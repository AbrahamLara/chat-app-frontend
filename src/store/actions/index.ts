import Vue from 'vue';
import { ActionTree } from 'vuex';
import { RootState } from '@/store/store-states';
import { THEME } from '@/utils/theme-utils';
import { loginUser, registerUser } from '@/api/auth-api';
import { namespaceAlerts } from '@/store/modules';
import { createAlertMessage, isFormErrorMessage } from '@/utils/alerts-utils';
import { LoginFormFields, RegisterFormFields } from '@/utils/auth-utils';
import {
  ADD_ERROR,
  ADD_SUCCESS,
  SET_ERRORS,
} from '@/store/constants/alerts-constants';
import {
  SET_THEME,
  LOGIN_USER,
  REGISTER_USER,
  SET_IS_AUTHENTICATED,
} from '@/store/constants/root-constants';

// Actions that affect how the application looks and behaves.
const appActions: ActionTree<RootState, RootState> = {
  [SET_THEME]: ({ commit }, theme: THEME) => {
    commit(SET_THEME, theme);
  },
  [LOGIN_USER]: async ({ commit }, loginForm: LoginFormFields) => {
    try {
      const res = await loginUser(loginForm);
      const data = await res.json();

      if (res.ok) {
        // Store the provided token as a cookie.
        // Set the cookie to expire at the same time as the cookie.
        Vue.$cookies.set('token', data.token, '3m');
        // Set the user as authenticated.
        commit(SET_IS_AUTHENTICATED, true);
      } else if (isFormErrorMessage(data)) {
        commit(namespaceAlerts(SET_ERRORS), data);
      } else {
        commit(namespaceAlerts(ADD_ERROR), data);
      }
    } catch (event) {
      const genericMessage = createAlertMessage(
        'An error occurred trying to sign you in!'
      );
      commit(namespaceAlerts(ADD_ERROR), genericMessage);
    }
  },
  [REGISTER_USER]: async ({ commit }, registerForm: RegisterFormFields) => {
    try {
      const res = await registerUser(registerForm);
      const data = await res.json();

      if (res.ok) {
        commit(namespaceAlerts(ADD_SUCCESS), data);
      } else if (isFormErrorMessage(data)) {
        commit(namespaceAlerts(SET_ERRORS), data);
      } else {
        commit(namespaceAlerts(ADD_ERROR), data);
      }
    } catch (event) {
      const genericMessage = createAlertMessage(
        'An error occurred trying to register you!'
      );
      commit(namespaceAlerts(ADD_ERROR), genericMessage);
    }
  },
  [SET_IS_AUTHENTICATED]: ({ commit }, isAuthenticated: boolean) => {
    commit(SET_IS_AUTHENTICATED, isAuthenticated);
  },
};

export default appActions;
