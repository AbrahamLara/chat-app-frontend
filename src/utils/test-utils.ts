import VueRouter, { RouteConfig } from 'vue-router';
import Vuex from 'vuex';
import { AppState } from '@/store/store-states';
import appMutations from '@/store/mutations';
import appActions from '@/store/actions';
import appModules from '@/store/modules';
import { DEFAULT_ROOT_STATE } from '@/store';
import { FormAlertMessage } from '@/utils/alerts-utils';
import { createLocalVue } from '@vue/test-utils';
import VueCookies from 'vue-cookies';
import Vue from 'vue';
import Vuetify from 'vuetify';

/**
 * Describes the available options for setting up a local vue instance for a test file.
 */
interface LocalVueSetupOptions {
  /**
   * Indicates that vuex plugin should be used.
   */
  useVuex?: boolean;

  /**
   * Indicates that vuetify plugin should be used.
   */
  useVuetify?: boolean;

  /**
   * Indicates that vue cookies plugin should be used.
   */
  useVueCookies?: boolean;

  /**
   * Indicates that vue router plugin should be used.
   */
  useVueRouter?: boolean;
}

/**
 * Resolves a vuetify warning of not being able to locate a data-app attribute on an element because vuetify
 * components should be wrapped by a v-app component.
 *
 * ```
 * Ex: [Vuetify] Unable to locate target [data-app] in "v-menu"
 * ```
 */
export function resolveVuetifyAppDataWarning() {
  const app = document.createElement('div');
  app.setAttribute('data-app', 'true');
  document.body.append(app);
}

/**
 * Create a Vuex store instance for testing.
 *
 * TODO: Reuse to create store in main.ts.
 *
 * @param vx The Vuex value
 * @param customState A state object that modifies the default state
 */
export function createStore(
  vx: typeof Vuex,
  customState: Partial<AppState> = {}
) {
  const mutations = appMutations;
  const actions = appActions;
  const modules = appModules;

  return new vx.Store({
    state: { ...DEFAULT_ROOT_STATE, ...customState },
    actions,
    mutations,
    modules,
  });
}

/**
 * Creates a VueRouter instance for testing.
 *
 * TODO: Reuse to create router in main.ts.
 *
 * @param vr The VueRouter value
 * @param routes The route config. If not provided, then a default config is provided.
 */
export function createRouter(
  vr: typeof VueRouter,
  routes: RouteConfig[] = [{ path: '/' }]
) {
  return new vr({ routes, mode: 'history' });
}

/**
 * Creates a form error message for testing.
 */
export function createFormAlertMessage(
  field: string,
  message: string
): FormAlertMessage {
  return { field, message };
}

/**
 * Creates a local vue instance with options to use any vue plugins.
 */
export function createLocalVueInstance(options: LocalVueSetupOptions = {}) {
  const { useVueCookies, useVuex, useVuetify, useVueRouter } = options;
  const localVue = createLocalVue();

  if (useVueCookies) {
    Vue.use(VueCookies);
  }
  if (useVuex) {
    localVue.use(Vuex);
  }
  if (useVuetify) {
    Vue.use(Vuetify);
  }
  if (useVueRouter) {
    localVue.use(VueRouter);
  }

  return localVue;
}
