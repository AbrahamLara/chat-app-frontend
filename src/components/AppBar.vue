<template>
  <v-app-bar :class="{ 'pa-sai': true, white: !isDarkMode }" app flat fixed>
    <router-link class="text-decoration-none" to="/" exact>
      <div
        :class="{
          'white--text text-h5 d-flex align-center': true,
          'grey--text text--darken-4': !isDarkMode,
        }"
      >
        ChatApp
      </div>
    </router-link>
    <v-spacer></v-spacer>
    <div class="d-flex align-center">
      <div
        id="logout-btn"
        v-if="showLogout"
        class="app-bar__btn text-center rounded pa-2 pl-2 pr-2"
        v-ripple
        @click="handleLogout"
      >
        Logout
      </div>
      <template v-else>
        <router-link
          to="/auth/signin"
          :class="{
            'text-decoration-none text--accent': true,
            'text--secondary': $route.path !== '/auth/signin',
          }"
        >
          <div
            id="signin-btn"
            class="app-bar__btn text-center rounded pa-2 pl-2 pr-2"
            @click="clearErrors"
            v-ripple
          >
            Sign In
          </div>
        </router-link>
        <router-link
          to="/auth/signup"
          :class="{
            'text-decoration-none text--accent': true,
            'text--secondary': $route.path !== '/auth/signup',
          }"
        >
          <div
            id="signup-btn"
            class="app-bar__btn text-center rounded pa-2 pl-2 pr-2"
            @click="clearErrors"
            v-ripple
          >
            Sign Up
          </div>
        </router-link>
      </template>
      <v-divider class="mx-2 my-auto" style="height: 25px" vertical></v-divider>
    </div>
    <settings-btn></settings-btn>
  </v-app-bar>
</template>

<script lang="ts">
import Vue from 'vue';
import { Component, PropSync } from 'vue-property-decorator';
import { THEME } from '@/utils/theme-utils';
import { namespaceAlerts } from '@/store/modules';
import { CLEAR_ALERTS } from '@/store/constants/alerts-constants';
import { Action } from '@/utils/decorators';
import { SET_IS_AUTHENTICATED } from '@/store/constants/root-constants';
import { getSettings } from '@/utils/dynamic-imports';

@Component({
  name: 'app-bar',
  components: {
    'settings-btn': getSettings,
  },
})
export default class AppBar extends Vue {
  @Action(SET_IS_AUTHENTICATED) authenticateUser!: Function;
  /**
   * Prop sync is used so that the app bar can handle determining if dark mode is in use by just receiving the theme as
   * it changes.
   */
  @PropSync('theme', { type: String, default: THEME.LIGHT })
  syncedTheme!: string;
  @PropSync('authenticated', { type: Boolean, default: false })
  showLogout!: boolean;

  /**
   * Determines if app-bar is flat, meaning it has not elevation.
   */
  isFlat = true;

  /**
   * The method to call to unsubscribe from a listener.
   */
  unsubscribe!: Function;

  get isDarkMode() {
    return this.syncedTheme === THEME.DARK;
  }

  /**
   * Clear existing alerts in state.
   */
  clearErrors() {
    this.$store.commit(namespaceAlerts(CLEAR_ALERTS));
  }

  /**
   * Handles the state of the page to log the user out.
   */
  handleLogout() {
    // Get rid of the token cookie.
    this.$cookies.remove('token');

    this.authenticateUser(false);

    // This is to prevent a NavigationDuplicated error if the current page is the home page. This shouldn't be needed
    // once there is a page to navigate the user to that isn't the home page and a navigation guard is added to the
    // home route.
    if (this.$route.path !== '/') {
      this.$router.push('/');
    }
  }
}
</script>

<style>
.app-bar__btn {
  user-select: none;
  width: 80px;
}
</style>
