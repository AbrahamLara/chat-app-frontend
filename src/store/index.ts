import Vue from 'vue';
import Vuex from 'vuex';
import appActions from '@/store/actions';
import appMutations from '@/store/mutations';
import appModules from '@/store/modules';
import { THEME } from '@/utils/theme-utils';
import { AppState } from '@/store/store-states';

Vue.use(Vuex);

export const DEFAULT_ROOT_STATE: AppState = {
  theme: THEME.LIGHT,
};

export default new Vuex.Store<AppState>({
  state: { ...DEFAULT_ROOT_STATE },
  mutations: appMutations,
  actions: appActions,
  modules: appModules,
});
