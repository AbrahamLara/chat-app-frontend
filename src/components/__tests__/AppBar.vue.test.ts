import { mount, Wrapper } from '@vue/test-utils';
import Vuex, { Store } from 'vuex';
import Vuetify from 'vuetify';
import AppBar from '@/components/AppBar.vue';
import { AppState } from '@/store/store-states';
import { namespaceAuth } from '@/store/modules';
import { ADD_ERROR, ADD_SUCCESS } from '@/store/constants/alerts-constants';
import { createAlertMessage } from '@/utils/alerts-utils';
import VueRouter from 'vue-router';
import { THEME } from '@/utils/theme-utils';
import {
  createLocalVueInstance,
  createRouter,
  createStore,
} from '@/utils/test-utils';
import { SET_IS_AUTHENTICATED } from '@/store/constants/auth-constants';

const localVue = createLocalVueInstance({
  useVuetify: true,
  useVueRouter: true,
  useVuex: true,
  useVueCookies: true,
});

async function expectAlertsToClearFor(
  btnId: string,
  wrapper: Wrapper<any>,
  store: Store<AppState>
) {
  const state = store.state as AppState;

  // Make sure there are no alerts by default.
  expect(state.auth?.alerts.errors.length).toEqual(0);
  expect(state.auth?.alerts.successes.length).toEqual(0);

  // Add error and success alerts to state.
  store.commit(namespaceAuth(ADD_ERROR), createAlertMessage('error1'));
  store.commit(namespaceAuth(ADD_SUCCESS), createAlertMessage('success1'));

  // Expect state to contain both newly added alerts.
  expect(state.auth?.alerts.errors.length).toEqual(1);
  expect(state.auth?.alerts.successes.length).toEqual(1);

  // Manually trigger the click event for the provided btn id.
  await wrapper.find(`#${btnId}`).trigger('click');

  // Expect there to be no alerts after triggering the click.
  expect(state.auth?.alerts.errors.length).toEqual(0);
  expect(state.auth?.alerts.successes.length).toEqual(0);
}

describe('AppBar.vue', () => {
  let wrapper: Wrapper<any>;
  let store: Store<AppState>;
  let router: VueRouter;

  beforeEach(() => {
    const vuetify = new Vuetify();
    store = createStore(Vuex);
    router = createRouter(VueRouter);

    wrapper = mount(AppBar, {
      localVue,
      vuetify,
      store,
      router,
    });
  });

  it('does not render dark mode classes', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders dark mode classes', async () => {
    await wrapper.setProps({ theme: THEME.DARK });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders the logout button', async () => {
    await wrapper.setProps({ authenticated: true });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('clears alerts when clicking sign in button', async () => {
    await expectAlertsToClearFor('signin-btn', wrapper, store);
  });

  it('clears alerts when clicking sign up button', async () => {
    await expectAlertsToClearFor('signup-btn', wrapper, store);
  });

  it('un-authenticates the user when clicking logout', async () => {
    // Make sure user is not authenticated by default.
    expect(store.state.auth?.isAuthenticated).toBeFalsy();

    // Manually set user as authenticated.
    store.commit(namespaceAuth(SET_IS_AUTHENTICATED), true);

    // Expect the isAuthenticated state to be true.
    expect(store.state.auth?.isAuthenticated).toBeTruthy();

    // Set the current route to '/test'
    await router.push('/test');

    // Create a cookie.
    await wrapper.vm.$cookies.set('token', 'cookie');

    // Make sure the cookie exists.
    const cookie1 = await wrapper.vm.$cookies.get('token');
    expect(cookie1).not.toBeNull();

    // Make the logout button visible by passing authenticated as true.
    await wrapper.setProps({ authenticated: true });

    // Manually trigger the logout button.
    await wrapper.find('#logout-btn').trigger('click');

    const cookie2 = await wrapper.vm.$cookies.get('token');
    // Expect the cookie to be null since it was removed after clicking logout.
    expect(cookie2).toBeNull();

    // Expect the user to no longer be authenticated after clicking logout.
    expect(store.state.auth?.isAuthenticated).toBeFalsy();

    // Expect the current route to be '/' after clicking logout.
    expect(router.currentRoute.path).toBe('/');
  });
});
