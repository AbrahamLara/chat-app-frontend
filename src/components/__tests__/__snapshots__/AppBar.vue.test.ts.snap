// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar.vue does not render dark mode classes 1`] = `
<header
  class="v-sheet theme--light v-toolbar v-toolbar--flat v-app-bar v-app-bar--fixed pa-sai white"
  style="height: 64px; margin-top: 0px; transform: translateY(0px); left: 0px; right: 0px;"
>
  <div
    class="v-toolbar__content"
    style="height: 64px;"
  >
    <a
      aria-current="page"
      class="text-decoration-none router-link-exact-active router-link-active"
      href="/"
    >
      <div
        class="white--text text-h5 d-flex align-center grey--text text--darken-4"
      >
        
      ChatApp
    
      </div>
    </a>
     
    <div
      class="spacer"
    />
     
    <div
      class="d-flex align-center"
    >
      <a
        class="text-decoration-none text--accent text--secondary"
        href="/auth/signin"
      >
        <div
          class="app-bar__btn text-center rounded pa-2 pl-2 pr-2"
          id="signin-btn"
        >
          
          Sign In
        
        </div>
      </a>
       
      <a
        class="text-decoration-none text--accent text--secondary"
        href="/auth/signup"
      >
        <div
          class="app-bar__btn text-center rounded pa-2 pl-2 pr-2"
          id="signup-btn"
        >
          
          Sign Up
        
        </div>
      </a>
       
      <hr
        aria-orientation="vertical"
        class="mx-2 my-auto v-divider v-divider--vertical theme--light"
        role="separator"
        style="height: 25px;"
      />
    </div>
     
    <div
      class="v-menu"
    >
      <!---->
    </div>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light elevation-0 v-size--default"
      id="settings-btn"
      role="button"
      style="font-size: 16px;"
      type="button"
    >
      <span
        class="v-btn__content"
      >
        <i
          aria-hidden="true"
          class="v-icon notranslate mdi mdi-cog-outline theme--light"
        />
      </span>
    </button>
  </div>
</header>
`;

exports[`AppBar.vue renders dark mode classes 1`] = `
<header
  class="v-sheet theme--light v-toolbar v-toolbar--flat v-app-bar v-app-bar--fixed pa-sai"
  style="height: 64px; margin-top: 0px; transform: translateY(0px); left: 0px; right: 0px;"
>
  <div
    class="v-toolbar__content"
    style="height: 64px;"
  >
    <a
      aria-current="page"
      class="text-decoration-none router-link-exact-active router-link-active"
      href="/"
    >
      <div
        class="white--text text-h5 d-flex align-center"
      >
        
      ChatApp
    
      </div>
    </a>
     
    <div
      class="spacer"
    />
     
    <div
      class="d-flex align-center"
    >
      <a
        class="text-decoration-none text--accent text--secondary"
        href="/auth/signin"
      >
        <div
          class="app-bar__btn text-center rounded pa-2 pl-2 pr-2"
          id="signin-btn"
        >
          
          Sign In
        
        </div>
      </a>
       
      <a
        class="text-decoration-none text--accent text--secondary"
        href="/auth/signup"
      >
        <div
          class="app-bar__btn text-center rounded pa-2 pl-2 pr-2"
          id="signup-btn"
        >
          
          Sign Up
        
        </div>
      </a>
       
      <hr
        aria-orientation="vertical"
        class="mx-2 my-auto v-divider v-divider--vertical theme--light"
        role="separator"
        style="height: 25px;"
      />
    </div>
     
    <div
      class="v-menu"
    >
      <!---->
    </div>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light elevation-0 v-size--default"
      id="settings-btn"
      role="button"
      style="font-size: 16px;"
      type="button"
    >
      <span
        class="v-btn__content"
      >
        <i
          aria-hidden="true"
          class="v-icon notranslate mdi mdi-cog-outline theme--light"
        />
      </span>
    </button>
  </div>
</header>
`;

exports[`AppBar.vue renders the logout button 1`] = `
<header
  class="v-sheet theme--light v-toolbar v-toolbar--flat v-app-bar v-app-bar--fixed pa-sai white"
  style="height: 64px; margin-top: 0px; transform: translateY(0px); left: 0px; right: 0px;"
>
  <div
    class="v-toolbar__content"
    style="height: 64px;"
  >
    <a
      aria-current="page"
      class="text-decoration-none router-link-exact-active router-link-active"
      href="/"
    >
      <div
        class="white--text text-h5 d-flex align-center grey--text text--darken-4"
      >
        
      ChatApp
    
      </div>
    </a>
     
    <div
      class="spacer"
    />
     
    <div
      class="d-flex align-center"
    >
      <div
        class="app-bar__btn text-center rounded pa-2 pl-2 pr-2"
        id="logout-btn"
      >
        
      Logout
    
      </div>
       
      <hr
        aria-orientation="vertical"
        class="mx-2 my-auto v-divider v-divider--vertical theme--light"
        role="separator"
        style="height: 25px;"
      />
    </div>
     
    <div
      class="v-menu"
    >
      <!---->
    </div>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light elevation-0 v-size--default"
      id="settings-btn"
      role="button"
      style="font-size: 16px;"
      type="button"
    >
      <span
        class="v-btn__content"
      >
        <i
          aria-hidden="true"
          class="v-icon notranslate mdi mdi-cog-outline theme--light"
        />
      </span>
    </button>
  </div>
</header>
`;
